snuba/migrations/context.py

```diff
@@ -4,7 +4,12 @@
 from snuba.migrations.status import Status
 
 
-class Context(NamedTuple):
+class MigrationContext(NamedTuple):
     migration_id: str
     logger: logging.Logger
     update_status: Callable[[Status], None]
+
+
+class OperationContext(NamedTuple):
+    single_node: bool
+    table_name: str
```

Comment:
The content of this operation context feels quite arbitrary.
I have a hard time finding how to define what it should contain and what it should not, which will make it hard to decide how it should evolve.
Single_node seems context that is not scoped to the single operation but the entire database. 
Table name seems very specific to create table operations.

So why don't you just pass the table name as a parameter to the engine and make it an attribute?

For single node. What seems you are trying to build here is an equivalent of a Strategy Pattern (https://en.wikipedia.org/wiki/Strategy_pattern).
Where you have the create table statement defined in abstract, then this relies on concrete strategies to build the table engine. Where you would have a strategy for single node and a strategy for multi node. 
The strategy is decided by the migration and passed as an object. The TableEngine class does not really encapsulate two different implementations, the right one has been decided earlier on. 

The benefit is that you separate two concerns that are not at the same abstraction level (distributed mode vs  all the other parameters to generate a table engine). Moreover you reduce the responsibiltiy  of the TableEngine in half, since now they would only have one path and rely on the strategy provided, which is likely to be noticeable when you are going to implement all of them.

I think building a simple version of this strategy contextualized to the TableEngine class where the concrete implementations only produce the engine type should be fairly easy and would allow you to drop this class. How does it sounds ?

Reply:
Also a pretty good signal that strategy pattern is what you are looking for is when you have multiple classes in your hierarchy extending the same class (directly or not) and in the same method all of them include the *same* condition like:
```
if a:
  do stuff
elif b:
  do other stuff
```

In your case this condition is the single_node one.

Reply:
The condition here only really is the single_node one which can be true or false and determines the SQL that gets generated. The only reason that table name is also passed here is to make specifying a create table operation less verbose and prone to error. I agree group those 2 in the context object is probably a bit confusing.

But if we required the user to pass the table name to the engine directly rather than forwarding it from the operation here, it means they have to specify it twice in a migration like `CreateTable(table_name="test", engine=MergeTree(table_name="test"))`. This makes using create table too verbose and prone to error as it raises the possibility the user passes a different value for the two fields which would be totally invalid.

Reply:
What if the CreateTable class passes it as parameter to MergeTree::get_sql ?

About the single_node approach I think the suggestion still stands.